text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,1389 @@
+  rpc ArrowStream(stream BatchArrowRecords) returns (stream BatchStatus) {}
+}
+```
+
+We use a stream-oriented protocol **to get rid of the overhead of specifying the schema and dictionaries for each
+batch.**
+A state will be maintained receiver side to keep track of the schemas and dictionaries.
+The [Arrow IPC format](#arrow-ipc-format) has been
+designed to follow this pattern and also allows the dictionaries to be sent incrementally.
+
+A `BatchArrowRecords` message is composed of 3 attributes. The protobuf definition is:
+
+```protobuf
+message BatchArrowRecords {
+  // [mandatory] Batch ID. Must be unique in the context of the stream.
+  string batch_id = 1;
+
+  // [mandatory] A collection of payloads containing the data of the batch.
+  repeated OtlpArrowPayload otlp_arrow_payloads = 2;
+
+  // [optional] Delivery type (BEST_EFFORT by default).
+  DeliveryType delivery_type = 3;
+}
+```
+
+The `batch_id` attribute is a unique identifier for the batch inside the scope of the current stream. It is used to
+uniquely
+identify the batch in the egress `BatchStatus` stream. See the [Batch Id generation](#batch-id-generation) section for
+more information on the implementation of this identifier.
+
+The `otlp_arrow_payloads` attribute is a list of `OtlpArrowPayload` messages. Each `OtlpArrowPayload` message represents
+a table of data encoded in a columnar format (metrics, logs, or traces). Although not used in the current version of
+this protocol, this attribute is an array to allow representing different types of entities.
+More details on the `OtlpArrowPayload` columns in the
+section [Mapping OTEL entities to Arrow records](#mapping-otel-entities-to-arrow-records).
+
+The `delivery_type` attribute is an optional attribute that can be used to indicate the type of message delivery
+guarantee expected by the sender. The delivery types are defined by the following protobuf enumeration:
+
+```protobuf
+enum DeliveryType {
+  BEST_EFFORT = 0;
+}
+```
+
+The delivery type `BEST_EFFORT` is the default value. This means that the sender expects the receiver to acknowledge
+receipt of a message and to do its best to pass the message on to the rest of the downstream chain.
+
+More specifically, an `OtlpArrowPayload` protobuf message is defined as:
+
+```protobuf
+message OtlpArrowPayload {
+  // [mandatory] A unique id assigned to a sub-stream of the batch sharing the same schema, and dictionaries.
+  string sub_stream_id = 1;
+
+  // [mandatory] Type of the OTLP Arrow payload.
+  OtlpArrowPayloadType type = 2;
+
+  // [optional] Serialized Arrow dictionaries
+  repeated EncodedData dictionaries = 3;
+
+  // [mandatory] Serialized Arrow Record Batch
+  // For a description of the Arrow IPC format see: https://arrow.apache.org/docs/format/Columnar.html#serialization-and-interprocess-communication-ipc
+  bytes record = 4;
+
+  // [mandatory]
+  CompressionMethod compression = 5;
```

Comment:
This attribute gives us the flexibility to use a compression method external to that offered by Arrow. For example, to implement the ZSTD dictionary optimization we discussed together. However, I think that the definition of the corresponding enum needs some improvements.

Reply:
Yea, I'd say it would also be good to ensure that the proposal documentation is explicit that this compression is referring to the entire payloads rather than the Arrow IPC data buffer level compression.

Reply:
Created https://github.com/f5/otel-arrow-adapter/issues/87 for tracking

Reply:
Do we do compression at this level and not at the gRPC level so that we can have ztsd dictionary optimization? Have you checked if gRPC compression makes impossible to optimize dictionaries?

Reply:
@zeroshade and @tigrannajaryan I removed the `compression` field from the protobuf message `OtlpArrowPayload` and introduced a configuration parameter at the OTLP Arrow API level. This way it is possible to enable zstd compression at the Arrow IPC level or at the global gRPC level (with the standard OTEL configuration). This intermediary level that existed was indeed unnecessary. 

I think that the compression optimization of the ZSTD dictionary could be implemented either at the IPC Arrow level or at the gRPC level. This optimization doesn't yet exist for Arrow IPC, and AFAIK this optimization doesn't exist for gRPC either. We could probably contribute either approach ourselves to improve scenarios with many small messages.

Code diff visible here: https://github.com/f5/otel-arrow-adapter/pull/93/files
OTEP diff visible here: https://github.com/open-telemetry/oteps/pull/171/commits/c78c78efd9aab18c83244f99e5815d506e9f932c